main_boaz.tex

```diff
@@ -0,0 +1,910 @@
+\begin{align*}
+A &\in \mathbb{R}^{\nset{foo}{..m}} \\
+B &\in \mathbb{R}^{\nset{foo}{..m},\nset{bar}{..n}} \\
+C &\in \mathbb{R}^{\nset{foo}{..m},\nset{bar}{..n}}
+\end{align*}
+then $(A \ndot{foo} B) \ndot{bar} C$ and $A \ndot{foo} (B \ndot{bar} C)$ don't even have the same shape.
+
+\subsubsection{Vectors to vectors}
+
+A very common example of a function from vectors to vectors is the softmax:
+\begin{equation*}
+  \nfun{foo}{softmax} A = \frac{\exp A}{\nsum{foo} \exp A} \approx \nmatrix{foo}{bar}{
+    0.731 & 0.002 & 0.953 \\
+    0.269 & 0.998 & 0.047
+  }.
+\end{equation*}
+
+And it's also very handy to have a function that renames an axis.
+
+\begin{equation*}
+\nmov{bar}{baz}{A} = \nmatrix{foo}{baz}{
+  3 & 1 & 4 \\
+  1 & 5 & 9
+}
+\end{equation*}
+
+We assume that this operation will use any \emph{casting} as needed. 
+So for example, we can use  $\name{widthtl}$ and $\name{heighttl}$  for width and height when the origin is at the top left corner, and 
+$\name{widthbl}$ and $\name{heightbl}$ for these axes when the origin is in the bottom left corner. 
+The renaming operation will automatically perform the appropriate casting. 
+
+
+
+Concatenation combines two vectors into one:
+\begin{align*}
+  A \ncat{foo} B &= \nmatrix{foo}{bar}{
+    3 & 1 & 4 \\
+    1 & 5 & 9 \\
+    2 & 7 & 1 \\
+    8 & 2 & 8
+  } \\
+  A \ncat{bar} B &= \nmatrix{foo}{bar}{
+    3 & 1 & 4 & 2 & 7 & 1 \\
+    1 & 5 & 9 & 8 & 2 & 8
+  }
+\end{align*}
+
+
+\subsubsection{Matrices}
+
+Finally, we briefly consider functions on matrices, for which you have to give \emph{two} axis names (and the order in general matters). Let $A$ be a named tensor with shape $\{\nset{foo}{..2},\nset{bar}{..2},\nset{baz}{..2\}}$:
+\begin{align*}
+A_{\nidx{foo}{1}} &= \nmatrix{bar}{baz}{
+  1 & 2 \\
+  3 & 4
+} \\
+A_{\nidx{foo}{2}} &= \nmatrix{bar}{baz}{
+  5 & 6 \\
+  7 & 8
+} \\
+\nfun{bar,baz}{det} A &= \nmatrix{}{foo}{\det \begin{bmatrix} 1 & 2 \\ 3 & 4 \end{bmatrix} & \det \begin{bmatrix} 5 & 6 \\ 7 & 8 \end{bmatrix}} \\
+\nfun{baz,bar}{det} A &= \nmatrix{}{foo}{\det \begin{bmatrix} 1 & 3 \\ 2 & 4 \end{bmatrix} & \det \begin{bmatrix} 5 & 7 \\ 6 & 8 \end{bmatrix}} \\
+\nfun{foo,bar}{det} A &= \nmatrix{}{baz}{\det \begin{bmatrix} 1 & 3 \\ 5 & 7 \end{bmatrix} & \det \begin{bmatrix} 2 & 4 \\ 6 & 8 \end{bmatrix}}
+\end{align*}
+For matrix inverses, there's no easy way to put a subscript under $\mathord\cdot^{-1}$, so we recommend writing $\nfun{foo,bar}{inv}$.
+
+\section{Examples}
+\label{sec:examples}
+
+\subsection{Attention}
+
+
+\begin{align*}
+  \text{Att} &\colon \mathbb{R}^{\name{seq},\name{key}} \times \mathbb{R}^{\name{seq},\name{key}} \times \mathbb{R}^{\name{seq},\name{val}} \rightarrow \mathbb{R}^{\name{seq},\name{val}} \\
+  \text{Att}(Q,K,V) &= \nfun{time}{softmax} \left( \frac{Q \ndot{key} K}{\sqrt{|\name{key}(Q)|}} \right) \ndot{seq} V
```

Comment:
This makes the product \ndot{key} align Q[seq] and K[seq], which is not right (the result type should still have a seq or seq' axis).

Reply:
You are right - should have here seq^in and  seq^out - will fix